feat(review): ask each fix for its test, and rule on non-convergence (#9596)
QwenLM/qwen-code: packages/core/src/skills/bundled/review/SKILL.md

4. **Failure scenario** — the concrete trigger and wrong outcome (for quality findings, the concrete cost or the quoted rule)
5. **Witness** — for a Critical: the observed output that settled the verdict, trimmed to the deciding lines — the probe's two sides, the A/B's quote pair, the sweep count over the real population, the failing test text — or the verifier's `not run — <reason>` line (Step 4's witness rule). A Suggestion carries one when a run produced it; it is not owed one.
6. **Suggested fix** — Concrete code suggestion when possible
7. **Fix witness** — the test that must go RED if that fix is removed (file + the behaviour it pins), or `N/A` when the fix adds no guard, branch or behaviour a test can pin. This is the ACCEPTANCE CRITERION for whoever fixes it, not the reviewer's evidence — `Witness` above is the evidence, and the two never substitute for each other.

> For **pattern-aggregated** findings, use the aggregated format from Step 4 (Pattern, Occurrences, Example, Failure scenario, Witness, Suggested fix, Fix witness, Severity) with the source tag added.

Author: **[Suggestion]** The diff updates only the reference line to name `Fix witness` in the pattern-aggregated format, but the format definition it points at — Step 4's "Pattern aggregation → 2. Format" list (SKILL.md lines 648-657) — was not updated and still ends at `Suggested fix`/`Severity` with no Fix witness field. The two halves of the doc now disagree. A reviewer aggregating a pattern (e.g. "missing error handling in 8 places") formats the finding per the Step 4 list, which never asks for the field, so the aggregate's artifact entry carries no `fixWitness`; Step 7's comment rule is keyed on that field, so none of the posted comments for the aggregate gets the acceptance criterion — yet multi-location aggregates are the highest-churn shape this PR exists to discipline.

```suggestion
Add a "**Fix witness:**" entry to the Step 4 aggregated-format list (after Suggested fix), mirroring the standalone item.
```

<details><summary>中文说明</summary>

本 diff 只更新了引用行，在模式聚合格式中列出 `Fix witness`，但它指向的定义——Step 4「模式聚合 → 2. 格式」清单（SKILL.md 648-657 行）——并未更新，仍止于 `Suggested fix`/`Severity`，没有 Fix witness 字段。文档的两半现在互相矛盾。聚合模式的审查者（如「8 处缺失错误处理」）按 Step 4 清单格式化 finding，而该清单从不要求该字段，于是聚合的 artifact 条目没有 `fixWitness`；Step 7 评论规则以该字段为键，聚合的所有已发布评论都没有验收标准——而多位置聚合恰恰是本 PR 要约束的返工率最高形态。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

- **still stands** — re-report it **under its original id**, updating the location if the code moved. It keeps its severity; a still-standing Critical blocks exactly as a new one would. Write that id into the re-report itself, immediately after the severity marker — `**[Critical]** R1-2: <the claim>` — and into the body entry if it cannot be anchored (`R1-2 <the claim>`). That prefix is not decoration: `compose-review` reads it back out of the comment when it builds the marker, and it is the only way an id survives into the machine ledger the next round recovers. Omit it and the same claim comes back renumbered, which is exactly what carrying the id forward exists to prevent.
- **cannot tell** — say so by id; a previous-round _Critical_ you cannot rule on joins `cannotTellCriticals` (it caps like any undecided blocker), a Suggestion is just disclosed.

> - **fix-induced** — the entry's own reported input is closed, but the change that closed it opened a new defect at the same site. Re-report the NEW defect **under the original id**, exactly as `still stands` prescribes (`**[Critical]** R1-2: <the new claim>`), with the new anchor and the new claim; the status line carries both facts — `R1-2 fix-induced — the round-2 fix closed the reported input and opened <new mechanism> at <file:line>; carried forward under R1-2`. See the fix-induced rule below for when this applies and when it must not.

Author: **[Critical]** The fix-induced rule allocates one id (the original entry's) per new defect but has no provision for two distinct fix-induced findings traced to the SAME previous entry — both re-report under the same id, and the duplicate-id refusal in `validateFindings` then fails the whole round. Concretely: a fix that introduces two distinct defects at one site (e.g. a guard that both drops a valid case and accepts an invalid one — both anchor in the same fix hunk, both trace to the same R1-2) makes the round's findings artifact carry two entries with id R1-2, and `validateFindings` throws `Duplicate finding id ... must be unique` (findings.ts:444-452, pinned by findings.test.ts:152-155) — a hard failure on the round's critical path that loses the entire round's compose, on exactly the churning-site PRs this rule targets. The guardrails mint a fresh id only when an operand is missing or the defect is weaker; neither holds for the second defect.

witness: not run — the duplicate-id throw is already pinned by findings.test.ts:152-155 against findings.ts:444-452; the half no run in this harness can settle is whether a live Step 6 agent writes two same-id fix-induced entries, which SKILL.md:798 explicitly instructs with no cap in the guardrails at 803-809.

```suggestion
Cap re-reports per original id at one per round: when the mechanical test would attribute a second new finding to an entry that already took a re-report this round, mint a fresh R<round>-<n> id (the unattributed direction the guardrail already favors) and count it in fresh but not induced.
```

<details><summary>中文说明</summary>

fix-induced 规则为每个新缺陷分配一个 id（原条目的 id），但没有任何条款处理「追溯到**同一条**上一轮条目的两个不同 fix-induced finding」——两者都会以同一 id 重新上报，`validateFindings` 的重复 id 拒绝随后使整轮失败。具体而言：一次修复在同一位置引入两个不同缺陷（如一个 guard 既丢弃合法用例又接受非法用例——两者锚定在同一修复 hunk，都追溯到同一个 R1-2）会使本轮 findings artifact 携带两条 id 均为 R1-2 的条目，`validateFindings` 抛出 `Duplicate finding id ... must be unique`（findings.ts:444-452，由 findings.test.ts:152-155 钉住）——这是轮次关键路径上的硬失败，会丢掉整轮 compose，恰恰发生在该规则针对的打转站点 PR 上。护栏只在操作数缺失或缺陷更弱时才铸造新 id；对第二个缺陷两者都不成立。

witness: not run —— 重复 id 抛出已被 findings.test.ts:152-155 对照 findings.ts:444-452 钉住；本环境无法裁决的另一半是「真实 Step 6 agent 是否会写出两条同 id 的 fix-induced 条目」——SKILL.md:798 在 803-809 的护栏无上限的情况下明确指示这样做。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

> - **fix-induced** — the entry's own reported input is closed, but the change that closed it opened a new defect at the same site. Re-report the NEW defect **under the original id**, exactly as `still stands` prescribes (`**[Critical]** R1-2: <the new claim>`), with the new anchor and the new claim; the status line carries both facts — `R1-2 fix-induced — the round-2 fix closed the reported input and opened <new mechanism> at <file:line>; carried forward under R1-2`. See the fix-induced rule below for when this applies and when it must not.

Author: **[Suggestion]** The new fix-induced rule never coordinates with the pre-existing bounded/unbounded-family rule in the same section: a new defect at a site under an open class-level finding is simultaneously told "fold it into the class id's re-report as evidence, never filed under its own id" (family rule) and — by the fix-induced rule — "re-report the NEW defect under the original id" (operand 2 satisfied) or "mint the fresh id" (guardrail 3). No precedence clause orders them, so the model's pick determines both the thread structure and the two census numbers. A model following the family rule leaves the folded sibling counting as neither `fresh` nor `induced`, so a genuinely churning round reads measured-and-low and resets the streak; a model following the fix-induced rule counts it `fresh`+`induced`, resurrecting the per-entrance enumeration the family rule exists to end. `induced` — the load-bearing input to the blocking non-convergence REQUEST_CHANGES — differs by the pick.

```suggestion
Add a precedence clause to the fix-induced rule: "a defect at a site whose family is already collapsed under a class id is not fix-induced — it folds into the class finding per the family rule and counts in neither fresh nor induced", or exempt collapsed-family sites in operand 2.
```

<details><summary>中文说明</summary>

新增的 fix-induced 规则与同节既有的有界/无界家族规则从未协调：位于已打开的类级 finding 站点上的新缺陷同时被告知「折叠进类 id 的再报中作为证据，绝不以自己的 id 上报」（家族规则）与——按 fix-induced 规则——「以原始 id 再报新缺陷」（操作数 2 满足）或「铸造新 id」（护栏 3）。没有优先级条款裁决二者，于是模型的选择同时决定线程结构与两个 census 数字。遵循家族规则的模型会让被折叠的兄弟既不进 `fresh` 也不进 `induced`，真正打转的轮次因此被判为「测得收敛」而清零 streak；遵循 fix-induced 规则的模型会把它计为 `fresh`+`induced`，复活家族规则要终结的逐入口枚举。`induced`——阻塞性非收敛 REQUEST_CHANGES 的关键输入——随选择而不同。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

> - **fix-induced** — the entry's own reported input is closed, but the change that closed it opened a new defect at the same site. Re-report the NEW defect **under the original id**, exactly as `still stands` prescribes (`**[Critical]** R1-2: <the new claim>`), with the new anchor and the new claim; the status line carries both facts — `R1-2 fix-induced — the round-2 fix closed the reported input and opened <new mechanism> at <file:line>; carried forward under R1-2`. See the fix-induced rule below for when this applies and when it must not.

Author: **[Suggestion]** The fix-induced disposition reuses the original id for a different claim, but the previous round's comment thread carrying that same id stays open on the PR forever — every rule in this section is written as if one id names one claim, and nothing tells the re-check how to rule a thread whose id's ledger entry now names a different defect. Concretely: round N posts "**[Critical]** R1-2: claim A"; round N+1's fix closes A and opens claim B at the same site, re-reported under R1-2, so the PR carries two live, id-tagged R1-2 threads. Round N+2's open-blocker re-check walks T1 (A's input closed → "fixed") and T2/the ledger (B — still open); the status table is keyed by id — one R1-2 row for two different rulings — and a model writing the R1-2 row "fixed" on T1's strength retires the standing blocker B without any ruling on it. Each further churn round adds another same-id thread, compounding precisely on the churning sites the mechanism targets.

```suggestion
State that from the fix-induced re-report round on, the id names the new claim only: the re-check rules the superseded thread "fixed" as a fact about the old claim, never against the ledger entry; or key thread rulings by claim content rather than by id alone.
```

<details><summary>中文说明</summary>

fix-induced 处置复用了原始 id 来表示一个**不同**的主张，但上一轮承载该 id 的评论线程会在 PR 上永远保持打开——本节的每条规则都按「一个 id 命名一个主张」来写，没有任何内容告诉再检查如何处理「其 id 的 ledger 条目现在命名了不同缺陷」的线程。具体而言：第 N 轮发布「**[Critical]** R1-2: 主张 A」；第 N+1 轮的修复关闭 A 并在同站打开主张 B，以 R1-2 再报，于是 PR 上有两条活跃的、带 id 标记的 R1-2 线程。第 N+2 轮的阻塞项再检查走 T1（A 的输入已关闭 →「fixed」）与 T2/ledger（B——仍打开）；状态表以 id 为键——一条 R1-2 行对应两种不同裁决——按 T1 的强度把 R1-2 行写成「fixed」的模型会在未裁决的情况下撤销活跃阻塞项 B。每多一轮打转就多一条同 id 线程，恰恰在该机制针对的打转站点上不断叠加。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

> - **fix-induced** — the entry's own reported input is closed, but the change that closed it opened a new defect at the same site. Re-report the NEW defect **under the original id**, exactly as `still stands` prescribes (`**[Critical]** R1-2: <the new claim>`), with the new anchor and the new claim; the status line carries both facts — `R1-2 fix-induced — the round-2 fix closed the reported input and opened <new mechanism> at <file:line>; carried forward under R1-2`. See the fix-induced rule below for when this applies and when it must not.

Author: **[Suggestion]** The fix-induced rule — disposition, two-operand test, and census — is written entirely for standalone findings and never addresses pattern aggregates, which are the review's highest-churn shape. For a fix-induced defect at one location of an aggregate, "the original id" is ambiguous (the finding's `R1-3` vs the per-location id `R1-3-2`), and the census unit is undefined (one finding vs per location). Concretely: round N posts an aggregate `R1-3` with 4 locations (each posting as its own suffixed-id comment); round N+1's fix closes location 2 and opens a new defect there — a model writing "**[Critical]** R1-3-2: <new claim>" is silently renumbered, because `LEDGER_ID_READBACK` (`/^(R\d+-\d+)[:.)\]]?(?=\s|$)/`, ledger.ts:191) cannot match `R1-3-2` (after `-3` the next char is `-`, not whitespace — probe-verified), so `buildLedger`'s `idFor` mints a fresh id — the exact renumbering the rule's "That prefix is not decoration" warning exists to prevent. A model that counts by finding and mints a fresh id undercounts `induced` (the bar needs two attributed findings; one miscounted site can flip the verdict). The code-age rule this block mirrors explicitly special-cases aggregates; the fix-induced rule does not.

```suggestion
Add an aggregate clause to the rule, mirroring the code-age rule's per-location treatment: state which id a fix-induced location re-reports under (the base id, satisfying LEDGER_ID_READBACK), and define the census unit for aggregates (one finding — one fresh/induced entry — regardless of how many locations are fix-induced).
```

<details><summary>中文说明</summary>

fix-induced 规则——处置、双操作数测试与 census——完全为独立 finding 而写，从未涉及模式聚合（review 中返工率最高的形态）。对于聚合某个位置上的 fix-induced 缺陷，「原始 id」是模糊的（finding 的 `R1-3` 与按位置的 `R1-3-2`），census 单位也未定义（一个 finding 还是按位置计数）。具体而言：第 N 轮发布带 4 个位置的聚合 `R1-3`（每个位置以各自带后缀 id 的评论发布）；第 N+1 轮的修复关闭位置 2 并在该处打开新缺陷——写「**[Critical]** R1-3-2: <新主张>」的模型会被静默重新编号，因为 `LEDGER_ID_READBACK`（`/^(R\d+-\d+)[:.)\]]?(?=\s|$)/`，ledger.ts:191）无法匹配 `R1-3-2`（`-3` 之后下一个字符是 `-` 而非空白——已探针验证），于是 `buildLedger` 的 `idFor` 铸造新 id——这正是规则「前缀不是装饰」警告要防止的重新编号。按 finding 计数并铸造新 id 的模型会低估 `induced`（门槛需要两条被归因的 finding；一处数错即可翻转裁决）。本块镜像的 code-age 规则明确为聚合单独设条款；fix-induced 规则没有。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

**The fix round is this loop's largest single source of its own next round — rule on that, do not just re-file it.** Measured across six multi-round pull requests, roughly a third of every post-first-round finding was introduced by the fix immediately preceding it (measured; DESIGN.md — The fix round that wrote the next round's findings (#9578)). Those findings are real and they post; what they must NOT do is arrive looking like independent new work, because a status table of eight fresh ids hides the fact that three of them are one site the loop has been circling. So before you mint `R<round>-<n>` for a finding, ask whether it is **fix-induced**, and take the disposition above when it is.

> **The test is mechanical on both operands, and both must hold.** (1) The finding's anchor falls inside a hunk **changed since the age reference** — the side file's `commitId`, validated and diffed exactly as the code-age rule below prescribes (`git --literal-pathspecs diff <commitId>..HEAD --unified=0 -- '<file>'`, same quoting, same pathspec proof, same two doubt states); code that predates the previous round cannot have been introduced by its fix. (2) A **previous-round ledger entry named that site** — the same file, and a line inside or adjacent to the hunk that answered it — and you can state the causal link in one clause: what the fix changed, and how that change produced this defect. A traced link, not an adjacency: two unrelated defects in one busy file are two findings.

Author: **[Suggestion]** The fix-induced mechanical test's operand 2 requires the previous entry to be in the same file as the new defect — "A previous-round ledger entry named that site — the same file" — so a defect the fix commit genuinely introduced in a different file (the classic "fix changed a function, broke its caller" shape) can never be attributed; it takes a fresh id and is excluded from `induced`, and the census's "honest measure" claim is false for that shape. Concretely: round 2's fix for R1-2 (anchored in a.ts) rewrites a.ts and updates its caller in b.ts, introducing a defect on a changed b.ts hunk. Operand 1 passes and the causal link is traceable, but operand 2 fails, so a compliant agent mints a fresh id; the round's census reads `fresh=N, induced=0` → measured-and-low → the streak resets. A PR whose fixes keep breaking callers in other files repeats this every round: `induced` stays 0 and the deterministic non-convergence block never fires on a genuinely churning PR. DESIGN.md's own provenance measurement (git-blame, file-agnostic) counted such spawns in its "~a third" baseline, so the rule's count systematically under-reads the measured phenomenon.

```suggestion
Broaden operand 2 to allow attribution when the previous entry is in a different file but the causal link is traced to the fix that answered it, or explicitly disclose in "What it buys" that induced measures same-site churn only.
```

<details><summary>中文说明</summary>

fix-induced 机械测试的操作数 2 要求上一轮条目与新缺陷**在同一文件**（「A **previous-round ledger entry named that site** — **the same file**」），因此修复提交真正在另一文件引入的缺陷（经典的「修了函数、弄坏了调用方」形态）永远无法被归因；它拿到新 id，被排除在 `induced` 之外，census「诚实度量」的主张对该形态为假。具体而言：第 2 轮对 R1-2（锚定在 a.ts）的修复重写 a.ts 并在同一次修复中更新其调用方 b.ts，在改动过的 b.ts hunk 上引入缺陷。操作数 1 通过、因果链可追溯，但操作数 2 失败，合规的 agent 铸造新 id；本轮 census 读作 `fresh=N, induced=0` → 测得偏低 → streak 清零。修复不断弄坏其他文件调用方的 PR 每轮重复此过程：`induced` 恒为 0，确定性的非收敛阻塞在真正打转的 PR 上永远不会触发。DESIGN.md 自身的溯源度量（git-blame，与文件无关）把这类衍生计入「约三分之一」基线，因此该规则的计数系统性地低估了实测现象。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

> **Three guardrails, and none of them is optional.** Attribution is a **bookkeeping** decision and never a posting one: a fix-induced finding posts, inline, at its own severity, exactly as it would under a fresh id — if you ever find yourself reaching for it to avoid reporting something, you have the rule backwards. It applies **only when the new defect is at least as severe and as confident as the entry it carries** — the same guard supersession carries, and for the same reason: a Critical id that quietly becomes a Suggestion retires a blocker nobody ruled on, so when the new defect is weaker, rule the entry `fixed` and file the new defect under its own fresh id. And when either operand is missing — no `commitId`, no worktree, the **context-unavailable** state, a previous entry you cannot identify, a causal link you cannot trace — **mint the fresh id**: unattributed is the safe direction, it is what every round did before this rule existed, and a wrong attribution is worse than none because it welds two claims to one id that later rounds cannot separate.

Author: **[Suggestion]** The severity/confidence guardrail forces a weaker fix-induced defect onto a fresh id ("rule the entry `fixed` and file the new defect under its own fresh id"), and the census contract counts only attributed findings in `induced` — so fix-induced defects the guardrail refused never enter `induced`, and the non-convergence rule can never fire on a loop whose fixes keep producing weaker defects, despite DESIGN.md's severity-agnostic premise ("roughly a third of every post-first-round finding was introduced by the fix immediately preceding it"). Concretely: a churning PR whose round-N fix closes a Critical entry and introduces a Suggestion-grade defect at the same site — the dominant unwitnessed-guard shape with a milder symptom. Operand 2 held and the causal link is traceable, but the guardrail refuses the id, the defect takes a fresh id, and the census measures `{fresh: ≥4, induced: 0}` — below the bar, so the streak resets. Round after round the loop keeps producing work the previous round created, and the deterministic non-convergence finding never fires: the mechanism's own "honest measure" reads zero on a loop that is entirely the previous round's product.

```suggestion
Count guardrail-refused fix-induced defects in induced anyway (operand 2 was satisfied; only the id was refused), or disclose in the census contract that induced counts same-or-stronger churn only, so the measured "third" is not read as severity-agnostic.
```

<details><summary>中文说明</summary>

严重度/置信度护栏迫使较弱的 fix-induced 缺陷走新 id（「把条目判为 `fixed`，以它自己的新 id 上报新缺陷」），而 census 契约只把**被归因**的 finding 计入 `induced`——于是被护栏拒绝的 fix-induced 缺陷永远不会进入 `induced`，非收敛规则在一个「修复不断产生较弱缺陷」的循环上永远不会触发，尽管 DESIGN.md 的前提与严重度无关（「首轮之后的 finding 中约三分之一由紧邻其前的修复引入」）。具体而言：打转 PR 的第 N 轮修复关闭一条 Critical 条目，并在同站引入一条 Suggestion 级缺陷——带较轻症状的主导形态「未见证 guard」。操作数 2 成立、因果链可追溯，但护栏拒绝该 id，缺陷拿新 id，census 测得 `{fresh: ≥4, induced: 0}`——低于门槛，streak 清零。一轮又一轮，循环持续制造上一轮创造的工作，而确定性的非收敛 finding 永不触发：该机制自称的「诚实度量」在一个完全是上一轮产物的循环上读数为零。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

**What it buys.** The ledger stops spending one id per round on a single churning site, so the marker's fifty-entry work list holds more distinct claims; the author reads one thread per site instead of a new one each round; and the count this produces — how many of the round's findings were fix-induced — is what the non-convergence rule below reads. That count is the honest measure of a loop's productivity, and it is not available to a review that renumbers everything every round.

> **Count the round as you rule it, and hand the two numbers over.** While you walk the findings above, keep a running census of exactly two numbers. **`fresh`** — how many findings FIRST APPEAR this round, counted over what the round REPORTS: the inline comments drafted for posting, the body Criticals, and the deferrals — the three channels `compose-review` cross-checks the number against: a `fresh` larger than everything reported across all three is refused as no census at all. The check is one-sided — an under-count passes it — so accuracy below that ceiling is yours to keep. Fix-induced findings count whether they took a previous id or a new one (they are new defects; the id is bookkeeping). NOT counted: the entries you ruled `still stands`, `fixed`, `cannot tell` or `superseded`, and findings that reach no channel — low-confidence findings are terminal-only, and a draft discarded as unanchorable posts nothing. Deferrals take `D<round>-<n>` ids in the artifact, but they ARE reports — count the ones MINTED this round: a deferral whose defect first appeared in an earlier round and is deferred again is not `fresh`, and counting re-deferrals every round inflates `fresh` with old never-induced findings, delaying the blocker precisely on the long-lived critical-floor pull requests this mechanism exists for. **`induced`** — how many of those `fresh` findings the fix-induced rule above **attributed**: the ones that TOOK a previous entry's id under that rule — the spent-id second defect of the guardrail above counts in `fresh` but not `induced`, exactly as it says there. `induced` is a SUBSET of `fresh` and can never exceed it. **It is the attributed count, not the count of findings on new lines**, and the difference is the whole precision of the mechanism: a pull request whose author pushed a new feature between rounds has most of its new findings on new lines and has NOT created them out of the review — there is no previous entry to trace them to, so they are `fresh` and not `induced`. A bar built on the looser number would block a pull request for growing. Carry the pair into the compose state as `convergence: {"fresh": N, "induced": M}` — one object, two integers, no prose. Omit the field entirely when you could not measure it: no `commitId`, no worktree, the **context-unavailable** state (the module refuses a census under it anyway, symmetric with round 1), or an age reference that failed validation. **Omitting is not the same as zero**, though both carry the streak: `compose-review` resets the streak only on a measured below-bar census with at least 4 `fresh` — absence, a malformed pair, and a census too small to be a trend (fewer than 4 `fresh`, zeros included) all carry the count untouched, because "could not measure" is not "measured and converging". Writing `{"fresh": 0, "induced": 0}` for a round you did not measure does not erase the standing claim — it stamps the marker's telemetry with a measured round that found nothing, a false trend point the next round reads as real. Omit the field: absence is the honest signal.

Author: Fixed in df6d20e6, but not by either of the two reconciliations — both change a count, and I think both counts are right.

The divergence you measured is real and it is a collision of WORDS, not of arithmetic. This mechanism counts defects the round newly identified; the convergence diagnosis counts inline comments posted for the first time. A fix-induced defect re-reported under the id it came from is new work by the first reading and a re-post by the second, so 6 and 2 are both correct about your probe's round. What was indefensible is that one body said "findings first filed in round 4" beside "2 of them reported for the first time" — two numbers under one phrase, and a reader can only conclude one of them is wrong.

So the blocker now reads "Of the 11 defects round 4 newly identified", and "first-appearing findings" became "newly identified defects". `isFreshDraft` and the marker's `fresh` are untouched.

On the two options you offered, for the record and for whoever edits this next — both are recorded as refusals in the module docblock and in SKILL: excluding carried-id re-reports from the census puts `induced` outside `fresh`, and `churnCensusOf` refuses `induced > fresh` as an impossible census, so every fix-induced round would silently stop counting; and marking them so `isFreshDraft` reads them as first-appearances tells the volume trend that a re-post is new work, which is the reading that function exists to refuse.

The remaining half of your finding I have NOT fixed: the persisted trend baseline does under-count carried-id induced findings, and that bias is real on churning pull requests. But `fresh` is #9461's definition of a posting-volume trend, where a re-post genuinely is not a first-time post, and re-defining it is outside this PR at round 5. Filing it as a follow-up rather than widening this diff.

Mutation-probed: reverting the blocker to the old vocabulary reds `compose-review.test.ts`.

> **Count the round as you rule it, and hand the two numbers over.** While you walk the findings above, keep a running census of exactly two numbers. **`fresh`** — how many findings FIRST APPEAR this round, counted over what the round REPORTS: the inline comments drafted for posting, the body Criticals, and the deferrals — the three channels `compose-review` cross-checks the number against: a `fresh` larger than everything reported across all three is refused as no census at all. The check is one-sided — an under-count passes it — so accuracy below that ceiling is yours to keep. Fix-induced findings count whether they took a previous id or a new one (they are new defects; the id is bookkeeping). NOT counted: the entries you ruled `still stands`, `fixed`, `cannot tell` or `superseded`, and findings that reach no channel — low-confidence findings are terminal-only, and a draft discarded as unanchorable posts nothing. Deferrals take `D<round>-<n>` ids in the artifact, but they ARE reports — count the ones MINTED this round: a deferral whose defect first appeared in an earlier round and is deferred again is not `fresh`, and counting re-deferrals every round inflates `fresh` with old never-induced findings, delaying the blocker precisely on the long-lived critical-floor pull requests this mechanism exists for. **`induced`** — how many of those `fresh` findings the fix-induced rule above **attributed**: the ones that TOOK a previous entry's id under that rule — the spent-id second defect of the guardrail above counts in `fresh` but not `induced`, exactly as it says there. `induced` is a SUBSET of `fresh` and can never exceed it. **It is the attributed count, not the count of findings on new lines**, and the difference is the whole precision of the mechanism: a pull request whose author pushed a new feature between rounds has most of its new findings on new lines and has NOT created them out of the review — there is no previous entry to trace them to, so they are `fresh` and not `induced`. A bar built on the looser number would block a pull request for growing. Carry the pair into the compose state as `convergence: {"fresh": N, "induced": M}` — one object, two integers, no prose. Omit the field entirely when you could not measure it: no `commitId`, no worktree, the **context-unavailable** state (the module refuses a census under it anyway, symmetric with round 1), or an age reference that failed validation. **Omitting is not the same as zero**, though both carry the streak: `compose-review` resets the streak only on a measured below-bar census with at least 4 `fresh` — absence, a malformed pair, and a census too small to be a trend (fewer than 4 `fresh`, zeros included) all carry the count untouched, because "could not measure" is not "measured and converging". Writing `{"fresh": 0, "induced": 0}` for a round you did not measure does not erase the standing claim — it stamps the marker's telemetry with a measured round that found nothing, a false trend point the next round reads as real. Omit the field: absence is the honest signal.

Author: Follow-up filed: #9674.

**One finding, one name.** A high-effort PR review also writes the incremental cache's cross-round `findings` ledger (Step 8), whose ids are `R<round>-<n>` — use those same ids here: a finding that will enter the ledger gets its `R<round>-<n>` as the artifact `id`, and a carried-forward finding keeps the id it already has. Two id schemes for one finding is how "R1-2" in next round's report and "f7" in this round's outcome ledger turn out to be the same defect that nobody can join. A finding the convergence posture deferred is still a confirmed finding and enters this artifact with all its fields — the deferral is a posting decision recorded in the compose state, never a severity change and never a reason to leave the artifact — but under its own id sequence, `D<round>-<n>`, **never consuming an `R<round>-<n>`**: the `R` counter must predict `buildLedger`, which numbers POSTED findings only, and a deferred finding holding `R6-2` would hand next round a ledger whose `R6-2` names a different defect than this round's artifact — the exact join "one finding, one name" exists to keep.

> Each entry carries `id` (unique — outcomes and resolved anchors both join on it), `severity`, `confidence`, `source`, `summary`, `failureScenario`, and either `file`/`line`/`anchor` or, for a pattern aggregate, a `locations[]` array with **one entry per location** (`suggestedFix`, `fixWitness`, `category`, `shortSummary` and `witness` are optional; `shortSummary` is derived from `summary` when absent; `witness` is the Step 4 witness — the executed evidence, or its `not run — <reason>` line — carried as data so the report and the comment bodies quote one recorded string instead of transcribing it twice more; `fixWitness` is the acceptance criterion the finding format asks for — the test that must go red if the suggested fix is removed, or `N/A` — carried for the same reason and read back by Step 7's comment body). The command validates the shape, refuses a duplicate id, refuses a finding with no failure scenario, sorts by severity → confidence → file → line → id, and writes counts nobody then recomputes by hand. Read the artifact for the numbers you quote in the Summary. This is a **canonicalization**, not a gate: it does not decide the verdict — `compose-review` does that, from the same findings — and it does not run at low effort, where the pass is unverified and emits no verdict.

Author: **[Suggestion]** The artifact-entry doc places `fixWitness` among the per-location optional fields of a pattern aggregate's `locations[]`, but the schema carries `fixWitness` only at the finding's top level: `FindingLocation` has just `file`/`line`/`anchor`, and `parseLocations` silently discards every other per-location field. Probe: a pattern aggregate with per-location `fixWitness` canonicalizes to `undefined`, while top-level `fixWitness` survives. A doc-compliant aggregate loses its acceptance criteria at canonicalization, and Step 7 posts the aggregate's per-location comments with no acceptance criterion — the mechanism silently does nothing for the pattern-aggregate shape (the highest-churn shape), with neither a runtime nor a type error to flag it.

```suggestion
Either add fixWitness to FindingLocation and round-trip it in parseLocations, or correct the SKILL.md to document fixWitness at the finding level (the aggregate carries one criterion for the whole finding), matching the schema the PR already ships.
```

<details><summary>中文说明</summary>

artifact 条目文档把 `fixWitness` 放在模式聚合 `locations[]` 的按位置可选字段里，但 schema 只在 finding 顶层携带 `fixWitness`：`FindingLocation` 只有 `file`/`line`/`anchor`，`parseLocations` 会静默丢弃其他一切按位置字段。探针：带按位置 `fixWitness` 的模式聚合规范化后为 `undefined`，而顶层 `fixWitness` 得以保留。符合文档的聚合在规范化时丢失验收标准，Step 7 为聚合各位置发布的评论没有验收标准——该机制对模式聚合形态（返工率最高形态）静默失效，且没有任何运行时或类型错误可以标记它。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

> Each entry carries `id` (unique — outcomes and resolved anchors both join on it), `severity`, `confidence`, `source`, `summary`, `failureScenario`, and either `file`/`line`/`anchor` or, for a pattern aggregate, a `locations[]` array with **one entry per location** (`suggestedFix`, `fixWitness`, `category`, `shortSummary` and `witness` are optional; `shortSummary` is derived from `summary` when absent; `witness` is the Step 4 witness — the executed evidence, or its `not run — <reason>` line — carried as data so the report and the comment bodies quote one recorded string instead of transcribing it twice more; `fixWitness` is the acceptance criterion the finding format asks for — the test that must go red if the suggested fix is removed, or `N/A` — carried for the same reason and read back by Step 7's comment body). The command validates the shape, refuses a duplicate id, refuses a finding with no failure scenario, sorts by severity → confidence → file → line → id, and writes counts nobody then recomputes by hand. Read the artifact for the numbers you quote in the Summary. This is a **canonicalization**, not a gate: it does not decide the verdict — `compose-review` does that, from the same findings — and it does not run at low effort, where the pass is unverified and emits no verdict.

Author: **[Suggestion]** One sentence of this diff promises `fixWitness` is carried so the report and comment bodies "quote one recorded string instead of transcribing it twice more" and are "read back by Step 7's comment body"; the other (line 1023) instructs the posted body to render it "in one sentence of ordinary prose" that must also "ask for the mutation that proves it". A criterion recorded in the format's terse shape ("file + the behaviour it pins") cannot satisfy the prose-sentence-and-mutation-ask rule without rewriting — so the two instructions cannot both hold, and the rewrite the artifact says the carry prevents is what the paragraph mandates. Step 7's composer follows the paragraph: the posted comment's acceptance sentence diverges from the recorded `fixWitness` string, so "read back by Step 7's comment body" and "quote one recorded string" are false by the doc's own instructions, and a later round comparing the posted comment to the recorded string sees a mismatch the docs themselves produced. An implementer fixing the readback gap by mechanically appending the recorded string still violates the "one sentence of ordinary prose" rule — the contradiction has no consistent resolution.

```suggestion
Align the two: either the comment quotes the recorded criterion verbatim (drop "in one sentence of ordinary prose... ask for the mutation" or fold it into the field's own format so the recorded string already contains the mutation ask), or correct the artifact line to say the criterion travels via the composer's rendering, not as a verbatim read-back.
```

<details><summary>中文说明</summary>

本 diff 的一句话承诺 `fixWitness` 被携带是为了让报告与评论正文「引用同一条已记录字符串，而非转写两次」，并「被 Step 7 评论正文读回」；另一句（1023 行）则指示已发布正文以「一句普通散文」呈现它，且必须「请求证明它的 mutation」。以格式的简略形态（「文件 + 它钉住的行为」）记录的验收标准，若不改写就无法满足「散文句 + mutation 请求」规则——两条指令不可能同时成立，而 artifact 声称被携带机制防止的改写，恰恰是另一段落的指令。Step 7 组装者按段落行事：已发布评论的验收句偏离记录的 `fixWitness` 字符串，于是按文档自身指令「被 Step 7 评论正文读回」与「引用同一条已记录字符串」都为假，后续轮次把已发布评论与记录字符串对比时会看到文档自己制造的错配。以机械追加记录字符串的方式修复读回缺口的实现者仍会违反「一句普通散文」规则——这个矛盾没有一致的解法。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

**A Critical's comment body carries its witness.** After the failure scenario, quote the observed output that settled the verdict — fenced, trimmed to the deciding lines — or the verifier's `witness: not run — <reason>` line (Step 4's witness rule). The witness is the difference between a comment the author can act on and a claim they have to re-derive before they can trust; the findings artifact already holds the string (`witness`), so this is a copy from data, not a fresh transcription.

> **And a comment whose fix adds a guard carries the test that must pin it.** When the finding's `fixWitness` is anything other than `N/A`, the posted body closes with it, in one sentence of ordinary prose: name the test that must fail if the fix is removed, and ask for the mutation that proves it (remove the guard, run that test, confirm it reds). One sentence, after the suggestion block — not a heading, not a checklist. This is the reviewer-side half of a measured loop: roughly a third of every post-first-round finding on six multi-round pull requests was introduced by the fix immediately before it, overwhelmingly as a guard or branch with no test of its own, and the deterministic gate re-runs only the tests that exist — so an unwitnessed guard passes every gate and returns as next round's finding. A fixer who is told the acceptance criterion closes it in THIS round; one who is not, does not (measured; DESIGN.md — The fix round that wrote the next round's findings (#9578)). The line reaches every fixer — a contributor, a maintainer, any bot — which is the point: the review cannot assume the fix comes from something it can configure. A finding whose `fixWitness` is `N/A` adds nothing (do not write "no test needed" — silence says it), and this sentence never changes what the comment reports or at what severity.

Author: **[Suggestion]** The new fix-witness paragraph claims the acceptance criterion "reaches every fixer", but the deferral channel defined in the same compose-state doc drops it: deferred findings post no comment, the body renders only `file:line — [source] title` (`DeferredEntry` in compose-review.ts carries `{file, line?, source, severity, title, locations?}` — no fixWitness — and `renderDeferredEntry` renders only that line), and the findings artifact that "enters with all its fields" is a local file, never a PR surface. Concretely: a PR under `auto` past round 6 (resolved `critical` floor) — a high-confidence Suggestion whose suggested fix adds a guard carries a non-`N/A` fixWitness in the artifact but is deferred under the posture; the author acts on the disclosed deferral list (the only PR surface that names the finding), fixes it without the test that must pin the guard, ships the unwitnessed guard — which passes every deterministic gate — and it returns as next round's finding, the exact cost the new paragraph says the mechanism eliminates.

```suggestion
State in the fix-witness paragraph (or the deferredSuggestions contract) that a deferred finding's criterion is not published — e.g., append the non-N/A fixWitness to the deferred body entry when it fits the 240-char cap, or scope the "reaches every fixer" claim to posted findings.
```

<details><summary>中文说明</summary>

新增的 fix-witness 段落声称验收标准「触达每一位修复者」，但同一 compose-state 文档定义的 deferral 通道会丢弃它：被延迟的 finding 不发布评论，正文只渲染 `file:line — [source] title`（compose-review.ts 的 `DeferredEntry` 只携带 `{file, line?, source, severity, title, locations?}`——没有 fixWitness——`renderDeferredEntry` 也只渲染那一行），而「以全部字段进入」的 findings artifact 是本地文件，从不构成 PR 表面。具体而言：`auto` 门限在第 6 轮后解析为 `critical` 的 PR——建议修复新增 guard 的高置信 Suggestion 在 artifact 中携带非 `N/A` fixWitness，却被姿态机制延迟；作者依据公开的延迟清单（唯一点名该 finding 的 PR 表面）行动，在没有必须钉住 guard 的测试的情况下修复，交付未见证的 guard——它能通过一切确定性 gate——下一轮作为 finding 回归，正是新段落声称该机制消除的成本。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

- `deferredSuggestions` — the findings the convergence posture deferred, as **typed entries** `{file, line?, source, severity, title, locations?}` copied from the findings artifact (Step 6's posture section — **high-confidence Suggestions that would otherwise post**, never low-confidence or Nice-to-have entries, which stay terminal-only; a `Critical` entry is relocated into the body Criticals, a malformed or free-text entry is refused). Deferred findings are **not** drafted into `comments` and are **not** counted toward `S` — the body renders them as a disclosed, non-capping list (up to 20 entries × 240 chars, overflow counted; the full set lives in the findings artifact), so the deferral is on the PR record without regenerating a review round. Non-deterministic entries **do** count toward the verifier-delivery floor — a deferred claim still publishes — while `source: build|test|probe` entries are excluded by that field exactly as body Criticals are by their tag: they are pre-confirmed, no verifier ever exists for them, and demanding one would cap the verdict with a gap no repair can close. A deferral never withholds the ledger anchor.

> - `convergence` — this round's census from Step 6's fix-induced rule, as `{"fresh": N, "induced": M}`: how many findings first appeared this round, and how many of those the fix-induced rule attributed to a previous entry's fix (the ATTRIBUTED count, not the count of findings on newly pushed lines). Two integers, `induced <= fresh`; a malformed pair, a float, a negative, or a numerator larger than its denominator is read as no census at all. **Omit the field when the round could not measure it** — absence carries the churn streak forward, a measured low census resets it, and zeros written for an unmeasured round are the one input that silently retires a standing non-convergence claim. `compose-review` owns everything downstream: the bar (half or more of `fresh`, and at least 4 `fresh`), the streak it stamps into the marker as `churnRounds`, and the body Critical it files itself on the second consecutive round above the bar.

Author: **[Suggestion]** The census never defines how deferred findings count in `fresh`, and its two definitions disagree: Step 6's enumeration counts "the ones that took a new `R<round>-<n>` id, plus the fix-induced ones that took a previous id" — while a deferred finding takes a `D<round>-<n>` id by design, so it is in neither bucket — yet the compose-state contract defines the census as "how many findings first appeared this round", which includes them. Concretely: from round 6 the `auto` floor turns `critical` and defers every otherwise-postable high-confidence Suggestion; a churning PR whose new defects are suggestion-grade then yields a well-formed measured census of `{"fresh": 0, "induced": 0}` — read as "this round converged" — and the streak resets. The round that should have been the second consecutive above-the-bar round firing the blocking non-convergence Critical instead silently retires the standing claim, exactly on the tail rounds where the floor turned critical.

```suggestion
State explicitly that deferred findings count in fresh (and in induced when the fix-induced rule attributed them), or name that deferred churn is intended not to count and define the all-deferred round (e.g. omit the census — absence carries the streak, which is fail-safe).
```

<details><summary>中文说明</summary>

census 从未定义被延迟的 finding 如何计入 `fresh`，且它的两个定义互相矛盾：Step 6 的枚举只数「拿到新 `R<round>-<n>` id 的，加上拿到上一轮 id 的 fix-induced 条目」——而被延迟的 finding 按设计拿 `D<round>-<n>` id，两个桶都不属于——compose-state 契约却把 census 定义为「本轮首次出现的 finding 数」，其中包含它们。具体而言：从第 6 轮起 `auto` 门限解析为 `critical`，延迟一切本可发布的 Suggestion；新缺陷均为 Suggestion 级的打转 PR 因此得到一个结构完整的测得 census `{"fresh": 0, "induced": 0}`——被解读为「本轮已收敛」——streak 被清零。本该作为「连续第二个越过门槛的轮次」触发阻塞性非收敛 Critical 的轮次，反而悄悄撤销了既有主张——恰恰发生在门限转严的尾部轮次上。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_

> - `convergence` — this round's census from Step 6's fix-induced rule, as `{"fresh": N, "induced": M}`: how many findings first appeared this round, and how many of those the fix-induced rule attributed to a previous entry's fix (the ATTRIBUTED count, not the count of findings on newly pushed lines). Two integers, `induced <= fresh`; a malformed pair, a float, a negative, or a numerator larger than its denominator is read as no census at all. **Omit the field when the round could not measure it** — absence carries the churn streak forward, a measured low census resets it, and zeros written for an unmeasured round are the one input that silently retires a standing non-convergence claim. `compose-review` owns everything downstream: the bar (half or more of `fresh`, and at least 4 `fresh`), the streak it stamps into the marker as `churnRounds`, and the body Critical it files itself on the second consecutive round above the bar.

Author: **[Suggestion]** (4 of 4) The compose-state `convergence` paragraph carries the entire malformed-input contract — "a malformed pair, a float, a negative, or a numerator larger than its denominator is read as no census at all", "zeros written for an unmeasured round are the one input that silently retires a standing non-convergence claim", and "the bar (half or more of `fresh`, and at least 4 `fresh`)" — and none of it is pinned; the third test's census pins all resolve to the Step 6 census section, not this compose-state contract (the compose-state doc itself has zero assertions). An edit deleting the failure-semantics sentence stays green; the compose-state doc is the text the model reads when writing `compose.json`, and with the failure semantics gone a model unsure of its numbers over-omits the field, the streak then carries where the round was measurable-and-low (which would have reset it), and the blocking non-convergence Critical files one round later than the census contract intends.

```suggestion
Pin a fragment of the compose-state paragraph's unique content, e.g. expect(body).toContain('a numerator larger than its denominator is read as no census at all').
```

<details><summary>中文说明</summary>

（4/4）compose-state 的 `convergence` 段落承载了全部畸形输入契约——「畸形对、浮点、负数或分子大于分母的输入一律视为没有 census」「为未测量轮写下的零是唯一会静默撤销既有非收敛主张的输入」「门槛（`fresh` 的一半或更多，且至少 4 条 `fresh`）」——但其中任何一条都没有被钉住；第三个测试的 census 钉全部落在 Step 6 census 小节，而不是这个 compose-state 契约（compose-state 文档自身零断言）。删除失败语义句子的编辑保持全绿；compose-state 文档是模型写 `compose.json` 时读的文本，失败语义消失后，对数字不确定的模型会过度省略该字段，streak 于是在「本可测量且偏低（本应清零）」的轮次上被携带，阻塞性非收敛 Critical 比 census 契约预期晚一轮触发。

</details>

_— deepseek-v4-flash via Qwen Code /review (v0.21.11)_
